### DISCUSSION TITLE
Inconsistent defaultValue behavior after Form submit?

### DISCUSSION THREAD

**Discussion by xHomu** [remix-run/remix#4318] [Q&A] [answered]
Not sure whether this is a Remix bug, or web bug, or a bug in my code.

Whether or not an uncontrolled input defaultValue would update after the loader revalidates depends on user behavior:

1. If a user navigates to the form via an internal Link, the input value does not change with latest loaderData return.
![image](https://user-images.githubusercontent.com/84349818/193706149-4d107a50-0586-48c1-ac04-812959faf9c0.png)

2. If a user enters the form directly (or refresh the page first), the input changes based on the revalidated loaderData.
![image](https://user-images.githubusercontent.com/84349818/193706191-d7919f16-edc3-4e7a-af67-2f2be9cbd54b.png)

The edit page action adds [Edit] to the post title to illustrate the discrepancy.

---

I created a minimal reproduction case based on the indie stack:

* deployed: https://remix-form-revalidate.fly.dev , 
* repo: https://github.com/homumado/remix-form-revalidate/blob/main/app/routes/notes/%24noteId/edit.tsx

----

## Reproduction Steps

1. login with rachel@remix.run / racheliscool
2. Create a new post
3. Edit the post
4. Save the post, the title input field wouldn't update with "[Edit] ..." (but the sidebar does)
![image](https://user-images.githubusercontent.com/84349818/193700365-5df07bd8-dc6b-42c6-b350-8bd2e76ae7cb.png)

5. Press ctrl+F5, the title now updates to "[Edit] ..."
6. Save again, this time after the refresh, the input field updates "[Edit] [Edit] ...".

![image](https://user-images.githubusercontent.com/84349818/193700388-e183f011-74dc-4440-9a40-7054323c9473.png)

```
Remix 1.7.2
React 17.0.2
Window 10
Chrome 106
```

**Comment by xHomu** [comment, marked as answer]
After further investigation, the discrepancy depends on these two conditions

1. Whether User access the Edit page via internal Link or directly (via refresh).
2. Whether the input field has been focused or not.

If  (1) direct access and (2) input has not been edited, the input value would reflect the revalidated loaderData, otherwise defaultValue would be "sticky".

Can someone explain why that might be the case?

**Reply by sergiodxa** [reply]
Because React doesn't control the inputs, when the form is accessed directly it will use the correct default value but after going to another route React keeps the same inputs on the screen (to avoid reconstructing the DOM) and because of that the inputs values are gonna be left untouched instead of going to the default ones.

**Reply by xHomu** [reply]
Got it.

Just making sure my mental model has updated correctly:

The discrepancy has to do with when react renders the input. Once a defaultValue is set, react cannot unset it without unmounting the whole component (hence the typical key prop workaround). This is why whether user actually interacted with the input matters.

**Reply by sergiodxa** [reply]
That's correct
